From b720e8589311d13f65cac686e2dda8597d0aed4f Mon Sep 17 00:00:00 2001
From: ValdikSS <iam@valdikss.org.ru>
Date: Mon, 26 May 2025 16:49:38 +0300
Subject: [PATCH 06/33] Poppler imagetoraster: fix mono dithering of 100% black
 pixel

If the pixel if fully filled, do not apply dithering algorithm (pattern) to it.

- Prevents white holes in the text
- As seen in pdftoraster / cfOneBitLine bitmap.c
---
 cupsfilters/imagetoraster.c | 10 ++++++----
 1 file changed, 6 insertions(+), 4 deletions(-)

diff --git a/cupsfilters/imagetoraster.c b/cupsfilters/imagetoraster.c
index e2295862..950c98de 100644
--- a/cupsfilters/imagetoraster.c
+++ b/cupsfilters/imagetoraster.c
@@ -2918,8 +2918,9 @@ format_K(imagetoraster_doc_t *doc,
 
         for (x = xsize; x > 0; x --)
         {
-          if (*r0++ > dither[x & 15])
-            *ptr ^= bitmask;
+	  if (*r0 > dither[x & 15] || *r0 == 0xff)
+	    *ptr ^= bitmask;
+	  r0++;
 
           if (bitmask > 1)
 	    bitmask >>= 1;
@@ -4274,8 +4275,9 @@ format_w(imagetoraster_doc_t *doc,
 
         for (x = xsize; x > 0; x --)
         {
-          if (*r0++ > dither[x & 15])
-            *ptr ^= bitmask;
+	  if (*r0 > dither[x & 15] || *r0 == 0xff)
+	    *ptr ^= bitmask;
+	  r0++;
 
           if (bitmask > 1)
 	    bitmask >>= 1;
-- 
2.53.0

